cpp/src/arrow/util/compression.h

```diff
@@ -21,32 +21,112 @@
 #include <cstdint>
 #include <memory>
 
-#include "arrow/status.h"
 #include "arrow/util/visibility.h"
 
 namespace arrow {
 
+class Status;
+
 struct Compression {
   enum type { UNCOMPRESSED, SNAPPY, GZIP, BROTLI, ZSTD, LZ4, LZO };
 };
 
 namespace util {
 
+/// \brief Streaming compressor interface
+///
+class ARROW_EXPORT Compressor {
+ public:
+  virtual ~Compressor();
+
+  /// \brief Compress some input.
+  ///
+  /// If bytes_read is 0 on return, then a larger output buffer should be supplied.
+  virtual Status Compress(int64_t input_len, const uint8_t* input, int64_t output_len,
+                          uint8_t* output, int64_t* bytes_read,
+                          int64_t* bytes_written) = 0;
+
+  /// \brief Flush part of the compressed output.
```

Comment:
Why and when should the user call Flush?

Reply:
That's a good question. The end goal of this PR is to allow implementing https://issues.apache.org/jira/browse/ARROW-1019 (compressed input and output streams), where we'd like to expose a flush operation (for output streams).

(unfortunately the semantics of flushing may be compressor-dependent: in some cases, perhaps not _all_ output is flushed, or the output is not valid compressed data until you've called End())